File: C++/프로그래머스/2/17686. ［3차］ 파일명 정렬/［3차］ 파일명 정렬.cpp
#include <string>
#include <vector>
#include <unordered_map>
#include <algorithm>

using namespace std;

vector<string> solution(vector<string> files) {
    unordered_map<string, pair<string, int>> nameMap;
    string head, number;
    int index;
    for(auto& file : files) {
        index = 0;
        head = "";
        number = "";
        
        while(file[index] < '0' || '9' < file[index]) {
            head += tolower(file[index++]);
        }
        while(index < head.size() + 5 && '0' <= file[index] && file[index] <= '9') {
            number += tolower(file[index++]);
        }
        nameMap[file] = {head, stoi(number)};
    }
    
    stable_sort(files.begin(),files.end(), [&](string a, string b)->int {
        if(nameMap[a].first == nameMap[b].first) return nameMap[a].second < nameMap[b].second;
        return nameMap[a].first < nameMap[b].first;
    });
    
    return files;
}

File: Python3/프로그래머스/2/17686. ［3차］ 파일명 정렬/［3차］ 파일명 정렬.py
def solution(files):
    files_list = {}
    
    index, head, number = 0, "", ""
    for file in files:
        index, head, number = 0, "", ""
        while not file[index].isdigit(): #HEAD
            head += file[index]
            index += 1
        while index < len(file) and file[index].isdigit(): #NUMBER
            number += file[index]
            index += 1
        files_list[file] = {"head": head.lower(), "number": int(number)}
    
    return sorted(files, key= lambda x : (files_list[x]["head"], files_list[x]["number"]))